### DISCUSSION TITLE
Mutations Guidance

### DISCUSSION THREAD

**Discussion by shaunnbarron** [nuxt/nuxt#19381] [Questions]
There's plenty of documentation around fetching data, but not much around writing it. My intuition would lead me to think the best thing would be something like:
```
const { execute, pending } = useFetch("/api/test", { method: "POST", body: reactiveFormData, immediate: false })
```
You would then call `execute` on form submission and use `pending` to show loading indicators. Unfortunately, this doesn't work for a couple of reasons:
* `pending` is true from the start even when the call hasn't been initiated.
* Changes to the reactive form data cause the fetch to fire

With all the talk around mutations/actions with many of the other meta-frameworks, this seems like a gap in the docs. What is the current best practice for performing mutations in Nuxt? Currently, I'm wrapping `useFetch` in my own function and mutating my own `pending` ref manually. This feels a little primitive though.

**Comment by TheAlexLichter** [comment]
If you send data, you commonly want to use `$fetch` (e.g. when sending a form).

See https://github.com/nuxt/nuxt/discussions/18731, https://github.com/nuxt/nuxt/pull/19357

**Comment by targetlucked69** [comment]
@manniL is there a mutation feature somewhere? Instead of calling `execute`, something that will let you mutate the cached data?

https://github.com/nuxt/nuxt/discussions/24627

**Reply by TheAlexLichter** [reply]
@walmartwarlord following up in the linked discussion

**Comment by Besmerch123** [comment]
Don't understand why this discussion gained so little attention. The mutations in Nuxt (their absence in particular) is the main reason why I still keen to use Tanstack query, possibility to provide a payload to a mutation function is soooo convenient, no need for parameters overhead state. +1 to author. 

And yeah, in Nuxt 4 `pending` is not true by default anymore, but still execution function can not accept custom payload.

**Reply by cernymatej** [reply]
Agreed that `useFetch` isn't very ergonomic for certain types of POST requests, such as form submissions. Simply adding a way to override the options in the `execute` function doesn't quite solve the issue imo.

Do you have an example of the kind of API you'd like to see?

**Reply by Besmerch123** [reply]
IMO, the Tanstack `useMutation` is quite good in it

```ts
const {
  data,
  error,
  isError,
  isIdle,
  isPending,
  isPaused,
  isSuccess,
  failureCount,
  failureReason,
  mutate,
  mutateAsync,
  reset,
  status,
  submittedAt,
  variables,
} = useMutation(
  {
    mutationFn,
    gcTime,
    meta,
    mutationKey,
    networkMode,
    onError,
    onMutate,
    onSettled,
    onSuccess,
    retry,
    retryDelay,
    scope,
    throwOnError,
  },
  queryClient,
)

mutate(variables, {
  onError,
  onSettled,
  onSuccess,
})
```
https://tanstack.com/query/latest/docs/framework/vue/reference/useMutation

This can be a good reference, but for me personally possibility to provide a payload to the mutation function is a killer feature

**Reply by cernymatej** [reply]
I feel like this is a little too much for something that is a core part of a framework, like `useFetch` is. Wouldn't it be better for you to just use TanStack query then? 😄

**Reply by Besmerch123** [reply]
100%, as I said earlier - I will be happy with a possibility to provide a payload to the mutation function. 

And another thing - I don't think that this should be supported by existing composables, like `useFetch` or `useAsyncData`, those composables is a good tools for data fetching, not for mutation. Mutations functions don't need cache, for example. They can mutate global Nuxt query cache with their response, but this should be the developers responsibility how to handle it.

Thus, `useMutation` composable will be a better fit for this.

And my minimal dream API for this may something like

```ts
const {
  data,
  error,
  pending,
  mutate,
  status,
} = useMutation(
  mutationKey: string,
  mutationFn: (payload: TPayload) => Promise<TResponse>, 
  {
    onError,
    onMutate,
    onSettled,
    onSuccess,
  }
)

mutate(payload: TPayload, {
  onError,
  onSettled,
  onSuccess,
})
```

**Reply by cernymatej** [reply]
So are you suggesting a new composable then? I got a little confused because the original post was about `useFetch`. In that case, it would probably be better to create an issue with a feature request so that more people can see it and share their ideas.

**Reply by Besmerch123** [reply]
Originally I was looking if it's possible to provide a payload to an `execute/refresh` function from `useFetch/useAsyncData` function, this is how I found this conversation and was surprised that his discussion had so little attention.

But then, after some thought and a conversation here, I came to conclusion that adding this feature to existing composables may be mess and error prone. Simply because of different mental model between fetching and writing data.

So yes, now I'm suggesting a new composable. And yes, a feature request may be a proper way to ask about it. And maybe I will do that)
But looks like there is not so many interest in that topic, as we can see from this discussion popularity.
I even would like to contribute in this one

**Reply by cernymatej** [reply]
Contributions are always welcome! 😊 Though I think a more visible discussion _(in an issue)_ is probably a better start for something like this. Feel free to mention it in the issue that you are open to implementing it.
